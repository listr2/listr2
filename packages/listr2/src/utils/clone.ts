import rfdc from 'rfdc'

const clone = rfdc({ circles: true })

/**
 * Deep clones a object in the easiest manner.
 */
// TODO: swap this with structured clone whenever we migrate to node 18
export function cloneObject<T extends Record<PropertyKey, any>>(obj: T): T {
  return clone(obj)
}
