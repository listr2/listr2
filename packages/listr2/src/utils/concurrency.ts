// adapted from https://github.com/fabiospampinato/promise-concurrency-limiter
// with MIT license, since it is again ESM only
// we needed to stop on the first failing case unlike the original
export class Concurrency {
  promise: Promise<void>

  private resolve: () => void
  private reject: (error: Error) => void
  private concurrency: number
  private count: number
  private queue: Set<() => void>

  constructor (options: { concurrency: number }) {
    this.concurrency = options.concurrency
    this.count = 0
    this.queue = new Set()
    this.initPromise()
  }

  public add<T>(fn: () => Promise<T>): Promise<T> {
    if (this.count < this.concurrency) {
      return this.run(fn)
    }

    return new Promise<T>((resolve) => {
      const callback = (): void => resolve(this.run(fn))

      this.queue.add(callback)
    })
  }

  private initPromise (): void {
    const { promise, resolve, reject } = this.withResolvers<void>()

    this.promise = promise
    this.resolve = resolve
    this.reject = reject
  }

  private flush (): void {
    for (const callback of this.queue) {
      if (this.count >= this.concurrency) {
        break
      }

      this.queue.delete(callback)

      callback()
    }
  }

  private run<T>(fn: () => Promise<T>): Promise<T> {
    this.count++

    const promise = fn()

    const cleanup = (): void => {
      this.count--

      this.flush()

      if (this.count === 0) {
        this.resolve()
      }
    }

    promise.then(cleanup, (error) => {
      this.queue.clear()
      this.reject(error)
    })

    return promise
  }

  private withResolvers<T>(): { promise: Promise<T>, resolve: (value: T) => void, reject: (error: Error) => void } {
    let resolve: (value: T) => void
    let reject: (error: Error) => void

    const promise = new Promise<T>((res, rej) => {
      resolve = res
      reject = rej
    })

    return {
      promise,
      resolve: resolve!,
      reject: reject!
    }
  }
}
