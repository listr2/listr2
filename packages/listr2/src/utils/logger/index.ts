export * from './logger.constants'
export type * from './logger.interface'
export * from './logger'
