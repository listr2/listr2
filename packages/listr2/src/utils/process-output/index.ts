export type * from './process-output-buffer.interface'
export * from './process-output-buffer'
export * from './process-output-stream'
export * from './process-output'
export type * from './process-output.interface'
export * from './writable'
