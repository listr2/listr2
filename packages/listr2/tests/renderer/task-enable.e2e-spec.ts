import { Listr } from '@root'
import type { MockProcessOutput } from '@tests/utils'
import { expectProcessOutputToMatchSnapshot, mockProcessOutput, unmockProcessOutput } from '@tests/utils'

describe('default renderer: task-enable', () => {
  const output: MockProcessOutput = {} as MockProcessOutput

  process.stdout.isTTY = true

  beforeEach(async() => {
    mockProcessOutput(output)
  })

  afterEach(async() => {
    unmockProcessOutput(output)
    jest.clearAllMocks()
  })

  // JeLYvYj1f4ddSjgwkUp7Nr8hICHLdEsU
  it('should enable task with general context', async() => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: (ctx): void => {
            ctx.skip = true
          }
        },

        {
          title: 'This task will never execute.',
          enabled: (ctx): boolean => !ctx.skip,
          task: (): void => {}
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true }
      }
    ).run()

    expectProcessOutputToMatchSnapshot(output, 'JeLYvYj1f4ddSjgwkUp7Nr8hICHLdEsU')
  })

  // 2PeJ1GS5SrNdC7DXMv8nHAQNfire6UYB
  it('should enable task with general context in subtasks', async() => {
    await new Listr(
      [
        {
          title: 'This task will execute.',
          task: (ctx): void => {
            ctx.skip = true
          }
        },

        {
          title: 'This task will show subtasks.',
          task: (_, task): Listr =>
            task.newListr(
              [
                {
                  title: 'This task will execute.',
                  task: (): void => {}
                },

                {
                  title: 'This task will never execute.',
                  enabled: (ctx): boolean => !ctx.skip,
                  task: (): void => {}
                }
              ],
              { concurrent: true }
            )
        }
      ],
      {
        concurrent: false,
        rendererOptions: { lazy: true }
      }
    ).run()

    expectProcessOutputToMatchSnapshot(output, '2PeJ1GS5SrNdC7DXMv8nHAQNfire6UYB')
  })
})
