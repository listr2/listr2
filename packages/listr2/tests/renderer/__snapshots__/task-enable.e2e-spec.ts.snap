// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`default renderer: task-enable should enable task with general context in subtasks: 2PeJ1GS5SrNdC7DXMv8nHAQNfire6UYB-exit 1`] = `[]`;

exports[`default renderer: task-enable should enable task with general context in subtasks: 2PeJ1GS5SrNdC7DXMv8nHAQNfire6UYB-stderr 1`] = `[]`;

exports[`default renderer: task-enable should enable task with general context in subtasks: 2PeJ1GS5SrNdC7DXMv8nHAQNfire6UYB-stdout 1`] = `
[
  "[?25l",
  "‚óº This task will execute.
",
  "[2K[1A[2K[G‚óº This task will execute.
‚óº This task will show subtasks.
",
  "[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
‚óº This task will show subtasks.
",
  "[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚óº This task will show subtasks.
",
  "[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚ùØ This task will show subtasks.
",
  "[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚ùØ This task will show subtasks.
  ‚óº This task will execute.
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚ùØ This task will show subtasks.
  ‚ùØ This task will execute.
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚ùØ This task will show subtasks.
  ‚úî This task will execute.
",
  "[2K[1A[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚úî This task will show subtasks.
",
  "[2K[1A[2K[1A[2K[G",
  "‚úî This task will execute.
‚úî This task will show subtasks.
",
  "[?25h",
]
`;

exports[`default renderer: task-enable should enable task with general context: JeLYvYj1f4ddSjgwkUp7Nr8hICHLdEsU-exit 1`] = `[]`;

exports[`default renderer: task-enable should enable task with general context: JeLYvYj1f4ddSjgwkUp7Nr8hICHLdEsU-stderr 1`] = `[]`;

exports[`default renderer: task-enable should enable task with general context: JeLYvYj1f4ddSjgwkUp7Nr8hICHLdEsU-stdout 1`] = `
[
  "[?25l",
  "‚óº This task will execute.
",
  "[2K[1A[2K[G‚óº This task will execute.
‚óº This task will never execute.
",
  "[2K[1A[2K[1A[2K[G‚ùØ This task will execute.
‚óº This task will never execute.
",
  "[2K[1A[2K[1A[2K[G‚úî This task will execute.
‚óº This task will never execute.
",
  "[2K[1A[2K[1A[2K[G‚úî This task will execute.
",
  "[2K[1A[2K[G",
  "‚úî This task will execute.
",
  "[?25h",
]
`;
